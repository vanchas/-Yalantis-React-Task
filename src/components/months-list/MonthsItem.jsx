import React, { useEffect, useState } from 'react'
import colorSchema from '../../data/Colors'
import styles from './month.module.css'

export default function MonthsItem({ month, users, setUsersToShow }) {
  const [monthUsersState, setMonthUsersState] = useState([])
  const [monthColor, setMonthColor] = useState(null)

  const sortUsersByMonths = () => {
    const usersInMonth = users
      .filter(user => new Date(user.dob).getMonth() + 1 === month.number);

    setMonthUsersState(usersInMonth)
    monthColorHandler(usersInMonth.length)
  }

  const monthColorHandler = (usersLength) => {
    colorSchema.forEach(col => {
      if (usersLength >= col.min
        && usersLength <= col.max) {
        setMonthColor(col.color)
      }
    })
  }

  useEffect(() => {
    if (users) sortUsersByMonths()
  }, [users])

  return (
    <li className={styles.list_item}
      style={monthColor
        ? { backgroundColor: monthColor }
        : null}
      onMouseEnter={() => setUsersToShow(monthUsersState)}
      onMouseLeave={() => setUsersToShow(null)}
    >
      <span>
        {month.name} <small>({monthUsersState.length})</small>
      </span>
    </li>
  )
}
