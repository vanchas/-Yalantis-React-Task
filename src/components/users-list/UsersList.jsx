import React from 'react'
import UsersListItem from './UsersListItem'
import styles from './users.module.css'

export default function UsersList({ users }) {

  return (
    <>
      {users && users.length ?
        <ul className={styles.list}>
          {users.map((user, index) => (
            <UsersListItem user={user} key={user.id} index={index + 1} />
          ))}
        </ul>
        : users && !users.length
          ? <div className={styles.message}>Sorry, there is no one user</div>
          : null
      }
    </>
  )
}

