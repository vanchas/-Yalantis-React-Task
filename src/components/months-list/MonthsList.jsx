import React from 'react'
import MonthsItem from './MonthsItem'
import styles from './month.module.css'
import months from '../../data/Months'

export default function MonthsList({ setUsersToShow, users }) {
  return (
    <ul className={styles.list}>
      {months.map(mon => (
        <MonthsItem
          month={mon}
          key={mon.number}
          users={users}
          setUsersToShow={setUsersToShow}
        />
      ))}
    </ul>
  )
}
