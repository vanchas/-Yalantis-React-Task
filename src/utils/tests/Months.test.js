import React from 'react'
import { shallow } from 'enzyme'
import { shallowToJson } from 'enzyme-to-json'

import MonthsList from '../../components/months-list/MonthsList'
import MonthsItem from '../../components/months-list/MonthsItem'
import UsersList from '../../components/users-list/UsersList'


describe('Link', () => {
  it('should render correctly', () => {
    const month = { name: "January", number: 1 }
    const users = [
      { firstName: 'Mark', lastName: 'Tven', id: '18f7tsdfg8sdt' },
      { firstName: 'Ilon', lastName: 'Mask', id: '2gas7itd7ait7' }
    ]
    const output = shallow(
      <MonthsItem
        month={month}
        users={users}
      />
    )
    expect(shallowToJson(output)).toMatchSnapshot()
  })

  it('should render correctly', () => {
    const users = [
      { firstName: 'Mark', lastName: 'Tven', id: '18f7tsdfg8sdt' },
      { firstName: 'Ilon', lastName: 'Mask', id: '2gas7itd7ait7' }
    ]
    const output = shallow(
      <MonthsList
        users={users}
      />
    )
    expect(shallowToJson(output)).toMatchSnapshot()
  })

  it('should render users list after mouseover event on month item', () => {
    const month = { name: "January", number: 1 }
    const users = [
      { firstName: 'Mark', lastName: 'Tven', id: '18f7tsdfg8sdt' },
      { firstName: 'Ilon', lastName: 'Mask', id: '2gas7itd7ait7' }
    ]
    const usersList = shallow(<UsersList />)
    const output = shallow(
      <MonthsItem
        month={month}
        users={users}
      />)
    output.simulate('mouseover')
    expect(shallowToJson(usersList)).toMatchSnapshot()
  });
})