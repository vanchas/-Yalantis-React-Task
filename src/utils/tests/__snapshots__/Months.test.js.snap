// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Link should handle state changes 1`] = `<Fragment />`;

exports[`Link should render correctly 1`] = `
<li
  className="list_item"
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
  style={null}
>
  <span>
    January
     
    <small>
      (
      0
      )
    </small>
  </span>
</li>
`;

exports[`Link should render correctly 2`] = `
<ul
  className="list"
>
  <MonthsItem
    key="1"
    month={
      Object {
        "name": "January",
        "number": 1,
      }
    }
    users={
      Array [
        Object {
          "firstName": "Mark",
          "id": "18f7tsdfg8sdt",
          "lastName": "Tven",
        },
        Object {
          "firstName": "Ilon",
          "id": "2gas7itd7ait7",
          "lastName": "Mask",
        },
      ]
    }
  />
  <MonthsItem
    key="2"
    month={
      Object {
        "name": "February",
        "number": 2,
      }
    }
    users={
      Array [
        Object {
          "firstName": "Mark",
          "id": "18f7tsdfg8sdt",
          "lastName": "Tven",
        },
        Object {
          "firstName": "Ilon",
          "id": "2gas7itd7ait7",
          "lastName": "Mask",
        },
      ]
    }
  />
  <MonthsItem
    key="3"
    month={
      Object {
        "name": "March",
        "number": 3,
      }
    }
    users={
      Array [
        Object {
          "firstName": "Mark",
          "id": "18f7tsdfg8sdt",
          "lastName": "Tven",
        },
        Object {
          "firstName": "Ilon",
          "id": "2gas7itd7ait7",
          "lastName": "Mask",
        },
      ]
    }
  />
  <MonthsItem
    key="4"
    month={
      Object {
        "name": "April",
        "number": 4,
      }
    }
    users={
      Array [
        Object {
          "firstName": "Mark",
          "id": "18f7tsdfg8sdt",
          "lastName": "Tven",
        },
        Object {
          "firstName": "Ilon",
          "id": "2gas7itd7ait7",
          "lastName": "Mask",
        },
      ]
    }
  />
  <MonthsItem
    key="5"
    month={
      Object {
        "name": "May",
        "number": 5,
      }
    }
    users={
      Array [
        Object {
          "firstName": "Mark",
          "id": "18f7tsdfg8sdt",
          "lastName": "Tven",
        },
        Object {
          "firstName": "Ilon",
          "id": "2gas7itd7ait7",
          "lastName": "Mask",
        },
      ]
    }
  />
  <MonthsItem
    key="6"
    month={
      Object {
        "name": "June",
        "number": 6,
      }
    }
    users={
      Array [
        Object {
          "firstName": "Mark",
          "id": "18f7tsdfg8sdt",
          "lastName": "Tven",
        },
        Object {
          "firstName": "Ilon",
          "id": "2gas7itd7ait7",
          "lastName": "Mask",
        },
      ]
    }
  />
  <MonthsItem
    key="7"
    month={
      Object {
        "name": "July",
        "number": 7,
      }
    }
    users={
      Array [
        Object {
          "firstName": "Mark",
          "id": "18f7tsdfg8sdt",
          "lastName": "Tven",
        },
        Object {
          "firstName": "Ilon",
          "id": "2gas7itd7ait7",
          "lastName": "Mask",
        },
      ]
    }
  />
  <MonthsItem
    key="8"
    month={
      Object {
        "name": "August",
        "number": 8,
      }
    }
    users={
      Array [
        Object {
          "firstName": "Mark",
          "id": "18f7tsdfg8sdt",
          "lastName": "Tven",
        },
        Object {
          "firstName": "Ilon",
          "id": "2gas7itd7ait7",
          "lastName": "Mask",
        },
      ]
    }
  />
  <MonthsItem
    key="9"
    month={
      Object {
        "name": "September",
        "number": 9,
      }
    }
    users={
      Array [
        Object {
          "firstName": "Mark",
          "id": "18f7tsdfg8sdt",
          "lastName": "Tven",
        },
        Object {
          "firstName": "Ilon",
          "id": "2gas7itd7ait7",
          "lastName": "Mask",
        },
      ]
    }
  />
  <MonthsItem
    key="10"
    month={
      Object {
        "name": "October",
        "number": 10,
      }
    }
    users={
      Array [
        Object {
          "firstName": "Mark",
          "id": "18f7tsdfg8sdt",
          "lastName": "Tven",
        },
        Object {
          "firstName": "Ilon",
          "id": "2gas7itd7ait7",
          "lastName": "Mask",
        },
      ]
    }
  />
  <MonthsItem
    key="11"
    month={
      Object {
        "name": "November",
        "number": 11,
      }
    }
    users={
      Array [
        Object {
          "firstName": "Mark",
          "id": "18f7tsdfg8sdt",
          "lastName": "Tven",
        },
        Object {
          "firstName": "Ilon",
          "id": "2gas7itd7ait7",
          "lastName": "Mask",
        },
      ]
    }
  />
  <MonthsItem
    key="12"
    month={
      Object {
        "name": "December",
        "number": 12,
      }
    }
    users={
      Array [
        Object {
          "firstName": "Mark",
          "id": "18f7tsdfg8sdt",
          "lastName": "Tven",
        },
        Object {
          "firstName": "Ilon",
          "id": "2gas7itd7ait7",
          "lastName": "Mask",
        },
      ]
    }
  />
</ul>
`;
